Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift

```diff
@@ -0,0 +1,263 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the swift-kafka-gsoc open source project
+//
+// Copyright (c) 2022 Apple Inc. and the swift-kafka-gsoc project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of swift-kafka-gsoc project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Crdkafka
+import Logging
+import NIOConcurrencyHelpers
+import NIOCore
+
+/// A back-pressure aware polling system for managing the poll loop that polls `librdkafka` for new acknowledgements.
+final class KafkaBackPressurePollingSystem {
+    /// The element type for the system, representing either a successful ``KafkaAcknowledgedMessage`` or a ``KafkaAcknowledgedMessageError``.
+    typealias Element = Result<KafkaAcknowledgedMessage, KafkaAcknowledgedMessageError>
+    /// The producer type used in the system.
+    typealias Producer = NIOAsyncSequenceProducer<Element, NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark, KafkaBackPressurePollingSystem>
+
+    /// The state machine that manages the system's state transitions.
+    let stateMachineLock: NIOLockedValueBox<StateMachine>
+
+    /// Closure that takes care of polling `librdkafka` for new messages.
+    var pollClosure: (() -> Void)?
+    /// The ``NIOAsyncSequenceProducer.Source`` used for yielding the messages to the ``NIOAsyncSequenceProducer``.
+    var sequenceSource: Producer.Source? {
+        get {
+            self.stateMachineLock.withLockedValue { stateMachine in
+                return stateMachine.sequenceSource
+            }
+        }
+        set {
+            self.stateMachineLock.withLockedValue { stateMachine in
+                stateMachine.sequenceSource = newValue
+            }
+        }
+    }
+
+    /// A logger.
+    private let logger: Logger
+
+    /// Initializes the ``KafkaBackPressurePollingSystem``.
+    ///
+    /// - Parameter logger: The logger to be used for logging.
+    init(logger: Logger) {
+        self.stateMachineLock = NIOLockedValueBox(StateMachine())
+        self.logger = logger
+    }
+
+    /// Runs the poll loop with the specified poll interval.
+    ///
+    /// - Parameter pollInterval: The desired time interval between two consecutive polls.
+    /// - Returns: An awaitable task representing the execution of the poll loop.
+    func run(pollInterval: Duration) async {
+        actionLoop: while true {
+            let action = self.stateMachineLock.withLockedValue { $0.nextPollLoopAction() }
+
+            switch action {
+            case .pollAndSleep:
+                self.pollClosure?()
+                do {
+                    try await Task.sleep(for: pollInterval)
+                } catch {
+                    self.stateMachineLock.withLockedValue { $0.shutDown() }
```

Comment:
We probably have to call `source.finish()` here?

Reply:
`source.finish()` gets invoked by `shutDown()` :+1:

Reply:
Yes but that's inside the state machine and we want to have this outside the state machine! Since the state machine should not execute any side effects